File: src/workspaceforge/scaffold/workspace-manifest.ts
import path from "node:path";
import { readJsonIfExists, writeJson } from "../util/fs-utils.js";
import type { StackDetectionResult } from "../scan/stack-detector.js";
import type { MemoryBackendDetectionResult } from "../memory-backends/registry.js";

/** Bumped whenever the shape of workspace.json changes in a way partial-state detection must know about. */
export const WORKSPACE_MANIFEST_SCHEMA_VERSION = "1";

export interface AdapterManifestEntry {
  installed: boolean;
  hookSchemaVersion: string | null;
}

export interface WorkspaceManifest {
  manifestSchemaVersion: string;
  workspaceforgeVersion: string;
  createdAt: string;
  lastScanAt: string;
  stack: {
    language: StackDetectionResult["language"];
    packageManager: StackDetectionResult["packageManager"];
    packages: number;
  };
  existingConfig: {
    claudeMd: boolean;
    agentsMd: boolean;
    cursorRules: boolean;
    copilotInstructions: boolean;
  };
  memoryBackends: MemoryBackendDetectionResult[];
  context: {
    rootContextKb: number;
    modules: string[];
  };
  adapters: Record<string, AdapterManifestEntry>;
}

/** Required top-level keys a valid workspace.json must have. Used by partial-state detection. */
export const REQUIRED_MANIFEST_KEYS: Array<keyof WorkspaceManifest> = [
  "manifestSchemaVersion",
  "workspaceforgeVersion",
  "createdAt",
  "lastScanAt",
  "stack",
  "existingConfig",
  "memoryBackends",
  "context",
  "adapters",
];

export function manifestPath(workspaceDir: string): string {
  return path.join(workspaceDir, "workspace.json");
}

export async function readManifest(workspaceDir: string): Promise<WorkspaceManifest | null> {
  return readJsonIfExists<WorkspaceManifest>(manifestPath(workspaceDir));
}

export async function writeManifest(workspaceDir: string, manifest: WorkspaceManifest): Promise<void> {
  await writeJson(manifestPath(workspaceDir), manifest);
}

/**
 * True if the given parsed value has every key a valid manifest requires.
 * Does not validate value types deeply -- partial-state detection only needs
 * to know "does this look like an interrupted or hand-edited manifest,"
 * not perform full schema validation.
 */
export function isManifestShapeValid(value: unknown): value is WorkspaceManifest {
  if (typeof value !== "object" || value === null) {
    return false;
  }
  const record = value as Record<string, unknown>;
  return REQUIRED_MANIFEST_KEYS.every((key) => key in record);
}

File: python/src/agenticworkspace/scaffold/workspace_manifest.py
"""
workspace.json read/write plus the shape-validation check partial-state
detection relies on. Ported from
src/agenticworkspace/scaffold/workspace-manifest.ts.

The manifest is represented as a plain dict with the same camelCase keys the
TypeScript CLI writes/reads, on purpose: both distributions can operate
against the same .workspace/ directory in the same repo, and workspace.json
is the one file both CLIs read back, so its wire shape must stay identical
across languages rather than being translated to snake_case internally.
"""
from __future__ import annotations

import os
from typing import Any, Dict, List, Optional

from ..util.fs_utils import read_json_if_exists, write_json

# Bumped whenever the shape of workspace.json changes in a way partial-state
# detection must know about.
WORKSPACE_MANIFEST_SCHEMA_VERSION = "1"

WorkspaceManifest = Dict[str, Any]

# Required top-level keys a valid workspace.json must have. Used by
# partial-state detection.
REQUIRED_MANIFEST_KEYS: List[str] = [
    "manifestSchemaVersion",
    "agenticworkspaceVersion",
    "createdAt",
    "lastScanAt",
    "stack",
    "existingConfig",
    "memoryBackends",
    "context",
    "adapters",
]


def manifest_path(workspace_dir: str) -> str:
    return os.path.join(workspace_dir, "workspace.json")


def read_manifest(workspace_dir: str) -> Optional[WorkspaceManifest]:
    return read_json_if_exists(manifest_path(workspace_dir))


def write_manifest(workspace_dir: str, manifest: WorkspaceManifest) -> None:
    write_json(manifest_path(workspace_dir), manifest)


def is_manifest_shape_valid(value: Any) -> bool:
    """
    True if the given parsed value has every key a valid manifest requires.
    Does not validate value types deeply -- partial-state detection only
    needs to know "does this look like an interrupted or hand-edited
    manifest," not perform full schema validation.
    """
    if not isinstance(value, dict):
        return False
    return all(key in value for key in REQUIRED_MANIFEST_KEYS)
